[RFC] build: many makefile fixes

mcprat opened the pull request:
(commit messages not done yet)

this is a large collection of fixes for small issues in the build system that have been bothering me

includes a fix based on #3945

some of them are obvious and easy fixes, but others need some discussion and might be wrong because I might have misunderstood something

for example, what is the purpose of the QUIET variable?

mcprat committed:
build: improve usage of flock

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
build: improve svn download method

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
rules: add and use more strip and escape functions

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
build: allow var.% targets to bypass the prepare steps

the same thing was done for the other debug rule "val.%"

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
rules: add debug rule to print type of variable

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
verbose.mk: verify extra file descriptor before redirection

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
toplevel.mk: allow forced check of build prerequisites

running 'make prepare-mk' does nothing if the build prerequisite check
was just done already

let this rule actually do the check again,
and also refresh the target and package manifests

for example: this allows someone to
update the manifest if they are corrupted
or update the symlinks in staging_dir/host/bin

also, give the rule some aliases

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
toplevel.mk: add some user friendly output

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
build: add GNU readlink to build prerequisites

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
build: allow host command symlinks to update

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
Revert "build: cleanup possibly dangling Python 2 host symlink"

This reverts commit 3b68fb57c938af3948ae4c2da61501183fbef649.

SetupHostCommand recipe has been updated
to check for matching symlink instead of the existence of file
so this quick fix is no longer needed

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
build: drop support for Python 3.6

python 3.6 was declared EOL in December 2021

new versions of meson have already set the requirement

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
config: add options to override host system arch

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
config: move some definitions to shared makefiles

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
build: clean up default build recipes

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
toplevel.mk: use variable for make operand -s and prefix @

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
toplevel.mk: fix duplicate calling of prepare-tmpinfo

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
build: cleanup PHONY target prerequisites

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
build: add basic GNU coreutils to build prerequisites

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
toplevel.mk: separate V= from OPENWRT_VERBOSE

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
build: let ldconfig host symlink to true be relative

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
build: do not allow Werror flag in packages

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
build: add python3-config to build prerequisites

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
build: purge usage of QUIET variable

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
build: sync variables between debug.mk and verbose.mk

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mpratt14 commented:
also it seems like the special target `.NOTPARALLEL`  is not being used anywhere...

mpratt14 commented:
I see already on the automatic check that prereq fails for Macs...

just need to know what directory coreutils would be in on those systems...

nbd168 commented on include/verbose.mk:
```diff
@@ -30,12 +30,16 @@ ifeq ($(IS_TTY),1)
 
 define ERROR_MESSAGE
-  printf "$(_R)%s$(_N)\n" "$(1)" >&8
+  [ -L /proc/$$$$/fd/8 ] && \
```
This is not portable and will break on non-Linux systems (e.g. macOS)

mpratt14 replied:
does it have to be in the form:

`test [ expression ]`

in order to work on macs?

if I look at docs for POSIX and the `-L` operand is there....

nbd168 commented:
> also it seems like the special target `.NOTPARALLEL` is not being used anywhere...

From the make manual:

>    If .NOTPARALLEL is mentioned as a target, then this invocation of make will be run serially, even if the ‘-j’ option is given. Any recursively invoked make command will still run recipes in parallel (unless its makefile also contains this target). Any prerequisites on this target are ignored.

mpratt14 commented on include/package.mk:
```diff
@@ -323,10 +323,7 @@ Build/Install=$(if $(PKG_INSTALL),$(call Build/Install/Default,))
 Build/Dist=$(call Build/Dist/Default,)
 Build/DistCheck=$(call Build/DistCheck/Default,)
 
-.NOTPARALLEL:
```
at first I was moving it around, removal is a mistake

I guess I would just have to ask you how it works because i thought (possibly wrongly) that if a target has a full empty line under it then theres nothing in it

 every example I find with a quick search shows targets listed after the colon of special targets like they are prerequisites of a normal target

so if that's not true how does it work? does it include every target under it or just the first one under it?

nbd168 replied:
If it's placed anywhere within the makefile, it applies to the whole make context

mpratt14 commented on include/prereq-build.mk:
```diff
@@ -8,6 +8,30 @@ include $(INCLUDE_DIR)/prereq.mk
 
+# an easy way to check for GNU coreutils
+$(eval $(call SetupHostCommand,[,Please install GNU coreutils, \
+	/bin/[ --version | grep GNU, \
```
checking for "true" is necessary (for the ldconfig situation), the rest just make sense to also check for. also, the other checks do not print "please install coreutils" when probably some of the other tools come with that

mpratt14 commented:
> From the make manual:
> 
> >    If .NOTPARALLEL is mentioned as a target, then this invocation of make


"this invocation of make"

does that mean it can be anywhere in the makefile and apply to the entire makefile?

Ansuel commented:
Mhhh considering the size of this wonder if a good idea would be to split this to smaller pr and merge the simple one?

mcprat committed:
toplevel.mk: update SILENT special target

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
build: move mkhash to tools and set as prerequisite

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
build: set flock as a build prerequisite

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
toplevel.mk: reuse some target names

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
build: make all use of find POSIX friendly

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
build: add function to find binaries in and out of PATH

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
scripts/env: fix git init for old git

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
build: require a recent version of Git

as of this commit:
git 2.16 has no patch support for more than 3 years prior
git 2.17 was patched 15 months prior

however, debian stretch still uses 2.11
to make the grep simple, start at 2.10

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
tools/pkgconf: install symlink for original name

otherwise it "looks" like it was never built

also in theory, there might be a package
that does not honor our environment variables
and demands that pkgconf is in PATH

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
rules: add debug rule to return host symlink

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
scripts: purge or improve usage of command -v

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
build: check for either curl or wget

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
build: purge makefile usage of command -v

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
build: add some variables for Build/Prepare

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
build: require a more recent version of perl

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
image-commands: make fake squashfs endianness option usable

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
build: add SOURCE_DIR variables

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
treewide: use make variable for rm and mv

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
treewide: remove whitespace only lines

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
timestamp.pl: add options and cleanup syntax

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
build: improve usage of timestamp.pl

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
build: specifically require GNU version of utilities

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
busybox: use libc glob()

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
tools/cmake: silence some deprecation warnings

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mpratt14 commented:
I addressed all previous comments and fixed some mistakes

I added some commits and separated 10 commits which can be handled in another PR

now just need to cleanup whatever else might have gone wrong during a rebase, and solve the macos host command issue, and then this will be ready for review again

mpratt14 commented:
looks like macos host commands are good now on the github bot, and there are no paths in the test commands at all 👍🏼 

waiting until tomorrow to rebase again (the last commit being a fixup test)

Ansuel commented:
_meanwhile 5 days after_

mpratt14 commented:
haha sorry, I found another small issue, also want to try building all packages to test

Ansuel commented:
If you want I can run this under a GitHub action for all the targets and
see the result

Il Mar 5 Lug 2022, 02:32 Michael Pratt ***@***.***> ha
scritto:

> haha sorry, I found another small issue, also want to try building all
> packages to test
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openwrt/openwrt/pull/9975#issuecomment-1174475715>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE2ZMQVPGKZJVGHE5LGL6XTVSN7BFANCNFSM5XQRX5QQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

mpratt14 commented:
i think it's ready for review again, and after it's decided which ideas are acceptable and which are not then I will finish the commit messages

a github action for all targets would be nice, for some reason my machine fails to build x86, but the reason seems unrelated.

Ansuel commented on package/system/ca-certificates/Makefile:
```diff
@@ -41,10 +41,17 @@ endef
+	SBINDIR="/usr/sbin" \
+	CERTSDIR="usr/share/ca-certificates" \
+	LOCALCERTSDIR="usr/local/share/ca-certificates"
+
```
Mhhh commit description doesn't mention this change. Can you better describe this and provide better title / commit description?

mpratt14 replied:
I dropped this commit in favor of another plan for the ca-certificates package which I'm going to handle in a new PR

Ansuel commented on include/prereq-build.mk:
```diff
@@ -183,8 +183,8 @@ $(eval $(call FindHostCommand,*readlink,Please install GNU 'readlink', \
 	readlink --version | grep GNU, \
 	greadlink --version | grep GNU))
 
-$(eval $(call SetupHostCommand,perl,Please install Perl 5.x, \
```
why?

Ansuel commented on include/prereq-build.mk:
```diff
@@ -183,8 +183,8 @@ $(eval $(call FindHostCommand,*readlink,Please install GNU 'readlink', \
-$(eval $(call SetupHostCommand,perl,Please install Perl 5.x, \
-	perl --version | grep "perl.*v5"))
+$(eval $(call SetupHostCommand,perl,Please install Perl >= 5.20, \
+	perl --version | grep 'perl.*v5.[2-9][0-9]'))
```
why?

mpratt14 replied:
i feel we should require a version of perl thats not *ancient* due to security bugs and syntax differences over time, a good example out of many:

https://www.effectiveperlprogramming.com/2014/07/perl-5-20-introduces-keyvalue-slices/

Ansuel commented on include/prereq.mk:
```diff
@@ -68,9 +68,22 @@ endef
 define TestHostCommand
   define Require/$(1)
-	($(3)) >/dev/null 2>/dev/null
+	$(if $(8),($(8) || $(7) || $(6) || $(5) || $(4) || $(3)), \
```
Wonder if this can be changed to for loop like others function in this mk do

mpratt14 commented:
> i know it's asking a lot but in the current state this is a nightmare to review

it was also a nightmare to create this 😆 

I completely agree that splitting this into multiple PRs must be part of the process, and I started with the commits regarding `download.pl` so that it is consistent with the change to how flock is used in `rules.mk` here

I do hope to see a Github bot action for all targets before I split it further, if that is possible

let me know if that is ready and I'll rebase again

mpratt14 commented:
I saw the commit in master for kernel CI jobs so I rebased

Ansuel commented:
Thanks curious of the results

mpratt14 commented:
rebuilding tools on my host to take a look.....

mpratt14 commented:
issue was download related, since I had flock creating the download file before it was checked by download.pl. I forgot to put it back

mcprat committed:
build: add compile fixup recipes

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
toolchain: cleanup dependencies

Reorganize dependencies between toolchain targets to save a line.

Use the gcc/final/install target stampfile as the prerequisite
for the toolchain/compile target stampfile.
The previous file name is incorrect,
the stampfile name format was changed at some point
but was never updated here.
So as a side-effect, some build time is saved
when rebuilding after the toolchain is already built.

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
build: use shell variable for SCAN_COOKIE

SCAN_COOKIE is a process ID number used to mark temp files.
This makes verbose logs have unique lines in prereq for different jobs,
making it difficult to compare diffs of log files.

SCAN_COOKIE is already exported where it is defined in the makefiles,
so use the shell variable in lines where it is printed to logs.
This conceals the number from logs.

Prefer the parent process ID (top level make) to match the job ID seen
when make is run in the background.

In scripts/feeds, don't redefine it if already defined by make,
and add it to the list of vars passed to make, to ensure the same value is used.

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
sdk: fix sdk initialization

prepare.mk is not being installed in the right place.
We have to use glob instead of listing each file,
otherwise the relative path is not included.

Set prepare.mk as a prerequisite
for prereq-build target in sdk.

Do not include the download and temp directories
in the git init.

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
build: add build variant to subtarget print

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
build: use common name for build target tools

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
build: rework autoremove of packages

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
target: add tools target

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
depends.mk: fix rdep typo

Signed-off-by: Michael Pratt <mcpratt@pm.me>

ehem reviewed (changes requested):
Unfortunately there is a simple fundamental issue with this commit.  Everything non-Git is disappearing.  This might have been valuable 2-4 years ago, now not one package is retrieved from Subversion and updating this doesn't seem worthwhile.

ehem reviewed (approved):
Seems reasonable.

ehem reviewed (commented):
At this point 3.7 is out of support and 3.8 is near its end.  I suspect the `main` branch needs to require 3.10 or 3.11 to developers forward before the next release branch.

ehem commented on include/prereq-build.mk:
```diff
@@ -173,13 +173,27 @@ $(eval $(call SetupHostCommand,python,Please install Python >= 3.7, \
 	python3.7 -V 2>&1 | grep 'Python 3', \
 	python3 -V 2>&1 | grep -E 'Python 3\.([7-9]|10)\.?'))
 
+$(eval $(call SetupHostCommand,python-config,Please install python-dev tools >= 3.7, \
+	python3.10-config --help | grep config, \
+	python3.9-config --help | grep config, \
+	python3.8-config --help | grep config, \
+	python3.7-config --help | grep config, \
+	python3-config --configdir | grep -E '3\.([7-9]|10)\.?'))
+
```
Why the duplication here?

ehem commented on include/autotools.mk:
```diff
@@ -54,7 +54,7 @@ endef
 define patch_libtool
 	$(Q)(cd $(1); \
-		for lt in $$$$($$(STAGING_DIR_HOST)/bin/find . -name ltmain.sh); do \
+		for lt in $$$$(find . -name ltmain.sh); do \
```
Better approach would be to spread more use of `$(FIND)`.

mpratt14 commented:
@ehem Thanks for the comments, but a lot of what's here is outdated anyway, or in some cases, already implemented. There are a few individual things here that we should still do, but they may not be implemented well.

This branch is basically an experimental branch that I built images on daily almost 2 years ago, I don't plan to do anything with it except using it as a template for other PRs I plan to make in the future

ehem commented on rules.mk:
```diff
@@ -459,12 +462,27 @@ define file_copy
 	$(CP) $(1) $(2)
 endef
 
+# POSIX usage of find with -mindepth and -maxdepth
+# $(1) => set of paths to recurse
+# $(2) => conditional statement(s)
+# $(3) => mindepth #
+# $(4) => maxdepth #
+# $(5) => eval expression (-exec)
+# $(6) => if set, do not follow links
+define find_depth
+	( \
+		for path in $(if $(word 2,$(1)),$(wildcard $(1)),$(1)); do \
+			$(FIND) $(if $(6),,-L) $$path '(' '!' -path "$(strip $$path$(call dir_depth,$(4)))" -o -prune ')' $(2) '(' '!' -path "$(if $(3),$(strip $$path$(call dir_depth,$(3))),*)" -o $(if $(5),$(5),-print) ')' ; \
+		done ; \
+	)
+endef
+
```
`-mindepth` and `-maxdepth` might not be POSIX, but I'm pretty sure nearly all flavors of Unix added them.  While nominally correct, I disagree with this due to the uniform availability of the `find` operators.

ehem commented on include/host-build.mk:
```diff
@@ -206,7 +206,7 @@ ifndef DUMP
     ifneq ($(CONFIG_AUTOREMOVE),)
       host-compile:
-		$(FIND) $(HOST_BUILD_DIR) -mindepth 1 -maxdepth 1 -not '(' -type f -and -name '.*' -and -size 0 ')' | \
+		$$(call find_depth,$(HOST_BUILD_DIR),'!' '(' -type f -name '.*' -size 0 ')',1,1) | \
```
`-not` might not be POSIX, but I believe support is widespread.  I strong dislike use of `'` for single-characters, I strongly prefer `\!`, `\(`, and `\)` instead.  I will admit opinions may differ though.

ehem commented on include/package-ipkg.mk:
```diff
@@ -210,7 +210,7 @@ $(_endef)
 	-find $$(IDIR_$(1)) -name 'CVS' -o -name '.svn' -o -name '.#*' -o -name '*~'| $(XARGS) rm -rf
 	$(Q)( \
-		find $$(IDIR_$(1)) -name lib\*.so\* -or -name \*.ko | awk -F/ '{ print $$$$NF }'; \
+		find $$(IDIR_$(1)) -name lib\*.so\* -o -name \*.ko | awk -F/ '{ print $$$$NF }'; \
```
Support for `-or` appears pretty common.  I don't disagree with this, but I'm dubious of the value.

ehem commented on include/package.mk:
```diff
@@ -262,7 +262,7 @@ define Build/CoreTargets
     compile:
 		-touch -r $(PKG_BUILD_DIR)/.built $(PKG_BUILD_DIR)/.autoremove 2>/dev/null >/dev/null
-		$(FIND) $(PKG_BUILD_DIR) -mindepth 1 -maxdepth 1 -not '(' -type f -and -name '.*' -and -size 0 ')' -and -not -name '.pkgdir' | \
+		$$(call find_depth,$(PKG_BUILD_DIR),'!' '(' -type f -name '.*' -size 0 ')' ! -name '.pkgdir',1,1) | \
```
Same comments as above.  I disagree with removing `-mindepth` and `-maxdepth` and strongly prefer the use of backslash for escaping single-characters.

ehem commented on include/rootfs.mk:
```diff
@@ -47,7 +47,7 @@ TARGET_DIR_ORIG := $(TARGET_ROOTFS_DIR)/root.orig-$(BOARD)
 ifdef CONFIG_CLEAN_IPKG
   define clean_ipkg
-	-find $(1)/usr/lib/opkg/info -type f -and -not -name '*.control' -delete
+	-find $(1)/usr/lib/opkg/info -type f '!' -name '*.control' | $(XARGS) $(RM)
```
`-delete` is also available on FreeBSD, so this might be readily available by fiat.

ehem reviewed (changes requested):
Spreading the use `-print0` and `$(FIND)` seem rather more valuable than most of this.  While several of these may not be POSIX, a quick look suggested they are widely available and hence not worth targeting.

ehem reviewed (approved):
Low priority, but one of these small things to go after.

ehem reviewed (commented):
This seems to conflict with 5dedc361759692f29fc926550e0e50ac3c84b774.  Don't fix things to work for older versions if you're going to require newer versions.
